File: src/crawler/store.js
/** Persistence for crawl results, on top of the existing `pages` table / PageService (app/crawler/store.py). */

import { connect, getPool, runQuery } from '../db/pool.js';
import { ConflictError } from '../utils/errors.js';
import { urlKey } from '../utils/urls.js';
import { PageService } from '../db/queries/pages.js';

/**
 * Store interface used by the crawler (the in-memory fake in tests implements the same methods):
 *   existingUrls(siteId) -> Map(urlKey -> stored url)
 *   upsertPage(siteId, page) -> id
 *   updateStatus(siteId, url, { httpStatus, redirectUrl }) -> boolean
 *   rollback()
 *   siteLock(siteId, fn) -> result of fn (throws ConflictError CRAWL_IN_PROGRESS if held)
 */
export class PgCrawlStore {
  constructor(pool = getPool()) {
    this.pool = pool;
    this.pages = new PageService(pool);
  }

  async existingUrls(siteId) {
    const { rows } = await runQuery(this.pool, 'SELECT url FROM pages WHERE site_id = $1', [siteId]);
    return new Map(rows.map((r) => [urlKey(r.url), r.url]));
  }

  async upsertPage(siteId, page) {
    const [id] = await this.pages.upsertPages(siteId, [page]);
    return id;
  }

  /** Update only status/redirect of an existing page (keeps its content). True if found. */
  async updateStatus(siteId, url, { httpStatus, redirectUrl }) {
    const result = await runQuery(
      this.pool,
      `UPDATE pages SET http_status = $1, redirect_url = $2, updated_at = now()
        WHERE site_id = $3 AND url = $4`,
      [httpStatus, redirectUrl, siteId, url],
    );
    return result.rowCount > 0;
  }

  async rollback() {
    // Every statement autocommits on the pool; nothing to roll back.
  }

  /**
   * Session-level PostgreSQL advisory lock on a dedicated connection. Prevents two crawls of the
   * same site running concurrently (a pooled connection could be reused by other queries).
   */
  async siteLock(siteId, fn) {
    const client = await connect(this.pool);
    const key = `crawl:${siteId}`;
    try {
      const { rows } = await runQuery(client, 'SELECT pg_try_advisory_lock(hashtextextended($1, 0)) AS acquired', [key]);
      if (!rows[0].acquired) {
        throw new ConflictError('A crawl for this site is already running', { code: 'CRAWL_IN_PROGRESS' });
      }
      try {
        return await fn();
      } finally {
        await runQuery(client, 'SELECT pg_advisory_unlock(hashtextextended($1, 0))', [key]);
      }
    } finally {
      client.release();
    }
  }
}

File: app/crawler/store.py
"""Persistence for crawl results, on top of the existing `pages` table / `PageService`."""

from __future__ import annotations

import uuid
from collections.abc import Iterator
from contextlib import AbstractContextManager, contextmanager
from typing import Protocol

from sqlalchemy import Engine, func, select, text, update
from sqlalchemy.orm import Session

from app.core.exceptions import ConflictError
from app.core.urls import url_key
from app.pages.models import Page
from app.pages.schemas import PageUpsert
from app.pages.service import PageService


class CrawlStore(Protocol):
    def existing_urls(self, site_id: uuid.UUID) -> dict[str, str]:
        """url_key -> stored URL for every page of the site."""
        ...

    def upsert_page(self, site_id: uuid.UUID, page: PageUpsert) -> uuid.UUID: ...

    def update_status(
        self, site_id: uuid.UUID, url: str, *, http_status: int, redirect_url: str | None
    ) -> bool:
        """Update only status/redirect of an existing page (keeps its content). True if found."""
        ...

    def rollback(self) -> None: ...

    def site_lock(self, site_id: uuid.UUID) -> AbstractContextManager[None]: ...


class SqlAlchemyCrawlStore:
    def __init__(self, session: Session, engine: Engine | None = None) -> None:
        self._session = session
        self._pages = PageService(session)
        self._engine = engine

    def existing_urls(self, site_id: uuid.UUID) -> dict[str, str]:
        urls = self._session.scalars(select(Page.url).where(Page.site_id == site_id))
        return {url_key(u): u for u in urls}

    def upsert_page(self, site_id: uuid.UUID, page: PageUpsert) -> uuid.UUID:
        return self._pages.upsert_pages(site_id, [page])[0]

    def update_status(
        self, site_id: uuid.UUID, url: str, *, http_status: int, redirect_url: str | None
    ) -> bool:
        result = self._session.execute(
            update(Page)
            .where(Page.site_id == site_id, Page.url == url)
            .values(http_status=http_status, redirect_url=redirect_url, updated_at=func.now())
        )
        self._session.commit()
        return bool(getattr(result, "rowcount", 0))

    def rollback(self) -> None:
        self._session.rollback()

    @contextmanager
    def site_lock(self, site_id: uuid.UUID) -> Iterator[None]:
        """Session-level PostgreSQL advisory lock on a dedicated connection.

        Prevents two crawls of the same site running concurrently. A dedicated connection is
        required because the ORM session returns its connection to the pool on each commit.
        """
        if self._engine is None:
            yield
            return
        with self._engine.connect() as conn:
            key = f"crawl:{site_id}"
            acquired = conn.execute(
                text("SELECT pg_try_advisory_lock(hashtextextended(:k, 0))"), {"k": key}
            ).scalar()
            if not acquired:
                raise ConflictError(
                    "A crawl for this site is already running", code="CRAWL_IN_PROGRESS"
                )
            try:
                yield
            finally:
                conn.execute(text("SELECT pg_advisory_unlock(hashtextextended(:k, 0))"), {"k": key})
                conn.commit()
